src/Tasks/Unzip.cs

```diff
@@ -0,0 +1,193 @@
+        /// </summary>
+        /// <param name="sourceArchive">The <see cref="ZipArchive"/> containing the files to extract.</param>
+        /// <param name="destinationDirectory">The <see cref="DirectoryInfo"/> to extract files to.</param>
+        private void Extract(ZipArchive sourceArchive, DirectoryInfo destinationDirectory)
+        {
+            foreach (ZipArchiveEntry zipArchiveEntry in sourceArchive.Entries.TakeWhile(i => !_cancellationToken.IsCancellationRequested))
+            {
+                FileInfo destinationPath = new FileInfo(Path.Combine(destinationDirectory.FullName, zipArchiveEntry.FullName));
+
+                if (!destinationPath.FullName.StartsWith(destinationDirectory.FullName, StringComparison.OrdinalIgnoreCase))
+                {
+                    // ExtractToDirectory() throws an IOException for this but since we're extracting one file at a time
+                    // for logging and cancellation, we need to check for it ourselves.
+                    Log.LogErrorFromResources("Unzip.ErrorExtractingResultsInFilesOutsideDestination", destinationPath.FullName, destinationDirectory.FullName);
+                    continue;
+                }
+
+                if (ShouldSkipEntry(zipArchiveEntry, destinationPath))
+                {
+                    Log.LogMessageFromResources(MessageImportance.Low, "Unzip.DidNotUnzipBecauseOfFileMatch", zipArchiveEntry.FullName, destinationPath.FullName, nameof(SkipUnchangedFiles), "true");
+                    continue;
+                }
+
+                try
+                {
+                    destinationPath.Directory?.Create();
+                }
+                catch (Exception e)
+                {
+                    Log.LogErrorWithCodeFromResources("Unzip.ErrorCouldNotCreateDestinationDirectory", destinationPath.DirectoryName, e.Message);
+                    continue;
+                }
+
+                if (OverwriteReadOnlyFiles && destinationPath.Exists && destinationPath.IsReadOnly)
+                {
+                    try
+                    {
+                        destinationPath.IsReadOnly = false;
+                    }
+                    catch (Exception e)
+                    {
+                        Log.LogErrorWithCodeFromResources("Unzip.ErrorCouldNotMakeFileWriteable", zipArchiveEntry.FullName, destinationPath.FullName, e.Message);
+                        continue;
+                    }
+                }
+
+                try
+                {
+                    Log.LogMessageFromResources(MessageImportance.Normal, "Unzip.FileComment", zipArchiveEntry.FullName, destinationPath.FullName);
+
+                    using (Stream destination = File.Open(destinationPath.FullName, FileMode.Create, FileAccess.Write, FileShare.None))
+                    using (Stream stream = zipArchiveEntry.Open())
+                    {
+                        stream.CopyToAsync(destination, 81920, _cancellationToken.Token);
```

Comment:
Curious what happens if another file exists in the destination and is longer than the archive entry. Does this override or does it just copy over the source stream, leaving some trailing bytes from the destination? The CopyTo documentation is a bit unclear on this.

Reply:
Oh, did not notice the FileMode.Create. Nevermind.

---
In reply to: [188796551](https://github.com/Microsoft/msbuild/pull/3291#discussion_r188796551) [](ancestors = 188796551)